jvm-streaming/src/Language/Java/Streaming.hs

```diff
@@ -0,0 +1,121 @@
+-- | Expose Java iterators as streams from the
+-- <http://hackage.haskell.org/package/streaming streaming> package.
+
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE QuasiQuotes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+{-# OPTIONS_GHC -fno-warn-unused-top-binds #-}
+
+module Language.Java.Streaming () where
+
+import Control.Distributed.Closure.TH
+import Data.IORef (IORef, newIORef, readIORef, writeIORef)
+import Data.Int (Int64)
+import Data.Singletons (SomeSing(..))
+import Data.Word (Word8)
+import Foreign.Ptr (FunPtr, castFunPtrToPtr, freeHaskellFunPtr, ptrToIntPtr)
+import qualified Foreign.JNI as JNI
+import Foreign.JNI.Types (jnull)
+import Language.Java
+import Language.Java.Inline
+import Streaming (Stream, Of)
+import qualified Streaming.Prelude as Streaming
+import System.IO.Unsafe (unsafePerformIO)
+
+isPoppableStream :: IORef (Stream (Of a) IO ()) -> IO Word8
+isPoppableStream ref = do
+    stream <- readIORef ref
+    Streaming.uncons stream >>= \case
+      Nothing -> return $ fromIntegral $ fromEnum False
+      Just _ -> return $ fromIntegral $ fromEnum True
+
+popStream :: Reflect a ty => IORef (Stream (Of a) IO ()) -> IO (J ty)
+popStream ref = do
+    stream <- readIORef ref
+    Streaming.uncons stream >>= \case
+      Nothing -> do
+        exc :: J ('Class "java.util.NoSuchElementException") <- new []
+        JNI.throw exc
+        return jnull
+      Just (x, stream') -> do
+        writeIORef ref stream'
+        reflect x
+
+type JNIFun a = JNIEnv -> JObject -> IO a
+
+foreign import ccall "wrapper" wrapObjectFun
+  :: JNIFun (J ty) -> IO (FunPtr (JNIFun (J ty)))
+foreign import ccall "wrapper" wrapBooleanFun
+  :: JNIFun Word8 -> IO (FunPtr (JNIFun Word8))
+foreign import ccall "wrapper" wrapFreeHaskellFunPtr
+  :: (FunPtr a -> IO ()) -> IO (FunPtr (FunPtr a -> IO ()))
+
+freeHaskellFunPtrPtr :: FunPtr (FunPtr a -> IO ())
+{-# NOINLINE freeHaskellFunPtrPtr #-}
+freeHaskellFunPtrPtr = unsafePerformIO (wrapFreeHaskellFunPtr freeHaskellFunPtr)
+
+newIterator :: FunPtr a -> FunPtr b -> IO (J ('Iface "java.util.Iterator"))
+newIterator hasNextPtr nextPtr = do
+    let hasNextIntPtr, nextIntPtr :: Int64
+        hasNextIntPtr = fromIntegral $ ptrToIntPtr $ castFunPtrToPtr hasNextPtr
+        nextIntPtr = fromIntegral $ ptrToIntPtr $ castFunPtrToPtr nextPtr
```

Comment:
`castFunPtrToPtr` has a precondition:
```
Note: this is valid only on architectures where data and
function pointers range over the same set of addresses, and should
only be used for bindings to external libraries whose interface
already relies on this assumption.
```

That's the only reason why the sparkle PR used StablePtr's instead.